File: lumen-examples/minimind/src/trainer/ppo.rs
use std::sync::Arc;

use lumen_core::{FloatDType, IndexOp, IntTensor, NoGradGuard, Tensor, D};
use lumen_dataset::{DataLoader, PairBatcher};
use lumen_nn::{functional::LossReduction, optim::{AdamW, Optimizer}, Linear, Module};
use lumen_transformer::{ForCausalLM, Sampler};
use minimind::{dataset::RLAIDataset, model::{MiniMindCache, MiniMindForCausalLM}, tokenizer::{EncodeOptions, Tokenizer}};

fn main() {
    if let Err(e) = result_main() {
        eprintln!("{e:?}");
    }
}

fn load_sft_model() -> anyhow::Result<(MiniMindForCausalLM<f32>, Arc<Tokenizer>)> {
    todo!()
}

#[derive(Module, Clone)]
pub struct MiniMindForCausalLMWithCritic<T: FloatDType> {
    pub actor_model: MiniMindForCausalLM<T>,
    pub value_head: Linear<T>,
}

impl<T: FloatDType> MiniMindForCausalLMWithCritic<T> {
    pub fn new(actor_model: MiniMindForCausalLM<T>) -> anyhow::Result<Self> {
        let vocab_size = actor_model.config.vocab_size;
        let value_head = Linear::new(vocab_size, 1, false, None)?;
        Ok(Self { actor_model, value_head })
    }
    
    pub fn forward(&self, input_ids: impl Into<IntTensor>, start_pos: usize, cache: &mut MiniMindCache<T>) -> anyhow::Result<(Tensor<T>, Tensor<T>)> {
        // (b, s, hidden) => (b, s, vocab_size)
        let logits = self.actor_model.forward(input_ids, start_pos, cache)?;

        // // (b, s, vocab_size) => (b, s, 1) => (b, s)
        let values = self.value_head.forward(&logits)?.squeeze(D::Minus1)?;
        
        Ok((logits, values))
    }
}

fn result_main() -> anyhow::Result<()> {
    let (actor_model, tokenizer) = load_sft_model()?;
    let model = MiniMindForCausalLMWithCritic::new(actor_model)?;
    let mut old_model = model.actor_model.copy()?;
    let mut ref_model = model.actor_model.copy()?;
    ref_model.eval();
    ref_model.requires_grad(false);

    let dataset = RLAIDataset::new("./assets/cache/xxx", tokenizer.clone())?;
    let loader = DataLoader::new(dataset, PairBatcher::default(), BATCH_SIZE, true);

    const EPOCHS: usize = 1;
    const BATCH_SIZE: usize = 32;
    
    // cache 不进行 kv cache，只保存 rotary embed 的 sin/cos
    let mut optimizer = AdamW::new(model.params(), Default::default())?;
    let mut cache = MiniMindCache::new(false, &model.actor_model.config)?;
    for _ in 0..EPOCHS {
        for batch in loader.iter() {
            let (prompts, _answers) = batch?;
            // prompts: list[string] -> 每个句子都是不完整的，需要 LLM 续写答案
            // 使用 actor_model 对所有问题进行一次生成 (batch, seq_len)
            let gen_out = generate(&model.actor_model, &tokenizer, &prompts)?;

            // 使用标准的 reward 模型对结果打分，每个 batch（句子）一个得分
            let rewards = calculate_rewards(&gen_out, &tokenizer)?; // (batch,)

            // 计算 ref model 对这些完整句子的概率
            let ref_logp = {
                let _gurad = NoGradGuard::new();
                let ref_logits = ref_model.forward(&gen_out, 0, &mut cache)?; // (batch, seq, vocab_size)
                // 同上，计算整个句子的概率
                let softmaxed_ref_logits = lumen_nn::functional::log_softmax(&ref_logits, D::Minus1)?; // (batch, seq)
                let ref_logp = softmaxed_ref_logits.sum(1)?; // (batch,)
                
                ref_logp
            };

            // 计算 old model 对这些完整句子的概率
            let old_logp = {
                let _gurad = NoGradGuard::new();
                let old_logits = old_model.forward(&gen_out, 0, &mut cache)?; // (batch, seq, vocab_size)
                // 同上，计算整个句子的概率
                let softmaxed_old_logp = lumen_nn::functional::log_softmax(&old_logits, D::Minus1)?; // (batch, seq)
                let old_logp = softmaxed_old_logp.sum(1)?; // (batch,)
                
                old_logp
            };

            // On-policy 收集数据，Off-policy 多次更新
            for _ in 0..5 {
                // 虽然我们得到了 actor 模型对每个句子的回答，但每个句子每个 token 的概率我们没有统计，这里需要重新传播（有梯度的）
                // 同时计算每个 batch（句子）的整体概率（完整 tracject 的概率）
                // 这里由于 actor 和 critic 放在一起，所以同时也返回了 critic_model 打分（对生成的 tokens 打分）
                // logits: (batch, seq, vocab_size)
                // values: (batch, seq)
                let (logits, values_seq) = model.forward(&gen_out, 0, &mut cache)?;
                // gen_out 左 shift 1 个位，表示 gen out 预测输出下一个 token id 
                let labels = gen_out.index((.., 1..))?; // (batch, seq_len - 1)
                // 先对每个 token 的 logits 先进行 softmax 转为概率（0-1, 求和为 1），再对每个概率进行 log 操作
                let softmaxed_logits = lumen_nn::functional::log_softmax(&logits, D::Minus1)?;
                // 但这个 token 的预测输出还是有 vocab_size 个，而其实上这个 token 的输出我们已经确定：labels 中的值，我们可以从所有（vocab_size）个概率
                // 提取 label 对应位置 token 的概率，就是在 generate 过程中这个 token 选择了下一个 token 的概率，得到
                // (batch, seq)：表示根据之前 token 预测下一个 token 的概率
                let logp_tokens = softmaxed_logits.gather(labels.unsqueeze(D::Minus1)?, 2)?.squeeze(D::Minus1)?;
                // 将一个 batch（句子）的 log 概率求和，就相当于对原始概率 * 后 log（因为 softmaxed_logits 中的概率值都被 log 过了！）
                let actor_logp = logp_tokens.sum(1)?; // (batch,)

                /*
                    - rewards: 理解为环境给我奖励
                    - values: 利用神经网络预测当前这个状态的价值    
                    二者的差距就是 advantages
                    如果 rewards 超过 values：advantage > 0 说明这个句子超出预期，是好的，我们要增大这个句子生成的概率
                */
                // 只取出 values_seq 的最后一个 token 的 values
                let seq_len = values_seq.dim(2)?;
                let values = values_seq.index((.., seq_len - 1))?; // (batch,)
                let advantages = &rewards - &values; // (batch,)

                // 统计误差：KL + 优势 + reward 和 cir model 的误差

                // 1. KL：actor_logp 表示对 actor_model 每个句子，生成整个回答的总概率的 log，同理对 ref model
                // log 的相减 = / 后 log。也就是用 actor_model 的概率 / ref_model 的概率再 log
                let kl_ref = (&actor_logp - &ref_logp).mean_all()?;
                // actor_model 的概率 / ref_model，重要性采样
                let ratio = (&actor_logp - &old_logp).exp()?; // (batch,)

                // advantages > 0 的句子表示我们要加强这个句子的生成概率，而 ratio 的上面就是 actor_model 生成这个句子的概率。
                // 如果 advantages > 0，backward 就会来提高 actor_model 生成这个句子的概率。
                let surr1 = &advantages * &ratio; // (batch,)
                let surr2 = ratio.clamp(1.-0.1, 1.+0.1)? * &advantages; // 限制增长过大
                let policy_loss = surr1.minimum(surr2)?.neg()?;
                // 训练 crit model
                let value_loss = lumen_nn::functional::mse_loss(&rewards, &values, LossReduction::Mean)?;

                let loss = policy_loss + 0.1 * value_loss + 0.1 * kl_ref;
                let grads = loss.backward()?;

                optimizer.step(&grads)?;
            }

            // 更新 old_model
            old_model = model.actor_model.copy()?;
        }
    }

    Ok(())
}

fn calculate_rewards(gen_out: &Tensor<u32>, tokenizer: &Arc<Tokenizer>) -> anyhow::Result<Tensor<f32>> {
    todo!()
}

fn generate<T: FloatDType>(
    model: &MiniMindForCausalLM<T>, 
    tokenizer: &Arc<Tokenizer>,
    prompts: &[String]
) -> anyhow::Result<Tensor<u32>> 
{
    let _gurad = NoGradGuard::new();
    let sample = Sampler::chat_default();

    let mut result = vec![];
    for prompt in prompts {
        // TODO: 长度对齐、是否返回 input_ids 部分。。。
        let input_ids = tokenizer.encode(&prompt, EncodeOptions::default())?.get_ids().to_vec();
        let output = model.generate(&input_ids, 1024, Some(tokenizer.eos_token_id()), &sample)?;
        result.push(Tensor::new(output)?);
    }

    let enc = Tensor::stack(&result, 0)?;
    Ok(enc)
}

File: examples/minimind/src/trainer/ppo.py
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import DataLoader

# 假设参数
BATCH_SIZE = 8
SEQ_LEN = 128      # 包含 Prompt + Generated Answer 的总长度
VOCAB_SIZE = 32000
DEVICE = "cuda" if torch.cuda.is_available() else "cpu"

def get_log_probs(logits, labels):
    """
    计算序列中每个 token 的对数概率并求和得到整句话的 log_p
    logits: [B, S, V]
    labels: [B, S]
    """
    # 1. 预测的是下一个 token，所以将 logits 往右移位，labels 往左移位对齐
    # logits: [B, S-1, V]
    shift_logits = logits[:, :-1, :].contiguous()
    # labels: [B, S-1]
    shift_labels = labels[:, 1:].contiguous()
    
    # 2. 计算 log_softmax: [B, S-1, V]
    log_probs = F.log_softmax(shift_logits, dim=-1)
    
    # 3. 根据 labels 索引对应的概率: [B, S-1, 1]
    per_token_logps = torch.gather(log_probs, dim=2, index=shift_labels.unsqueeze(-1))
    
    # 4. 求和得到整句话的对数概率: [B]
    return per_token_logps.squeeze(-1).sum(dim=1)

def train_ppo_step():
    # --- 1. 模型初始化 ---
    # Actor: 策略网络; Critic: 价值网络; Ref: 参考网络 (冻结)
    actor = MyLLM().to(DEVICE)           # [B, S] -> [B, S, V]
    critic = MyCritic().to(DEVICE)       # [B, S] -> [B, S]
    ref_model = MyLLM().to(DEVICE).eval()
    old_actor = MyLLM().to(DEVICE).eval() # 用于 PPO 重要性采样
    
    optimizer = torch.optim.AdamW(
        list(actor.parameters()) + list(critic.parameters()), lr=1e-5
    )

    # 模拟输入 Prompts: [B, S_prompt]
    prompts = torch.randint(0, VOCAB_SIZE, (BATCH_SIZE, 64)).to(DEVICE)

    # --- 2. Rollout 采样阶段 ---
    with torch.no_grad():
        # 生成回答，拼接成完整序列
        # gen_seq: [B, S] (S = prompt_len + gen_len)
        gen_seq = actor.generate(prompts, max_new_tokens=64) 
        
        # 计算 Reward (由外部 Reward Model 打分)
        # rewards: [B]
        rewards = reward_model(gen_seq) 
        
        # 计算 Ref 模型和 Old 模型的概率 (基准)
        # ref_logp: [B], old_logp: [B]
        ref_logits = ref_model(gen_seq)
        ref_logp = get_log_probs(ref_logits, gen_seq)
        
        old_logits = old_actor(gen_seq)
        old_logp = get_log_probs(old_logits, gen_seq)

    # --- 3. PPO 更新阶段 (多次迭代) ---
    for _ in range(5):
        # A. 前向传播
        # current_logits: [B, S, V]
        # current_values_seq: [B, S]
        current_logits = actor(gen_seq)
        current_values_seq = critic(gen_seq)
        
        # B. 计算当前 Actor 的 log_p
        # actor_logp: [B]
        actor_logp = get_log_probs(current_logits, gen_seq)
        
        # C. 计算 Advantage (优势)
        # 取最后一个 token 的 value 作为句子的状态价值估计
        # values: [B]
        values = current_values_seq[:, -1]
        # advantages: [B]
        advantages = rewards - values.detach() 
        # 归一化优点 (可选但推荐)
        advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)

        # D. 计算 Policy Loss (PPO Clip)
        # ratio: [B] (重要性采样比率)
        ratio = torch.exp(actor_logp - old_logp)
        
        surr1 = ratio * advantages
        surr2 = torch.clamp(ratio, 1.0 - 0.2, 1.0 + 0.2) * advantages
        # policy_loss: 标量
        policy_loss = -torch.min(surr1, surr2).mean()

        # E. 计算 Value Loss (Critic)
        # 使 Critic 预测的值接近真实 Reward
        # value_loss: 标量
        value_loss = F.mse_loss(values, rewards)

        # F. 计算 KL Divergence Penalty
        # 惩罚 Actor 偏离 Ref 模型过远
        # kl_div: 标量
        kl_div = (actor_logp - ref_logp).mean()

        # G. 总损失
        total_loss = policy_loss + 0.1 * value_loss + 0.05 * kl_div

        # H. 反向传播
        optimizer.zero_grad()
        total_loss.backward()
        optimizer.step()

        print(f"Loss: {total_loss.item():.4f} | KL: {kl_div.item():.4f}")

    # 更新 Old Actor 参数
    old_actor.load_state_dict(actor.state_dict())

# --- 辅助模型结构示例 ---
class MyLLM(nn.Module):
    # 标准 Transformer 结构
    def forward(self, x): 
        # x: [B, S] -> logits: [B, S, V]
        pass

class MyCritic(nn.Module):
    # 将 LLM 的 Hidden State 映射到一个标量 Value
    def forward(self, x):
        # x: [B, S] -> values: [B, S]
        pass